src/thread.rs

```diff
@@ -104,10 +114,182 @@ enum ThreadJoinStatus {
+                    // so assume some progress has been made.
+                    assume_progress = true;
+                }
+                *watch = *record_iteration
+            });
+
+            // The watch set is different so assume progress has been made
+            if assume_progress {
+                log::trace!("Atomic watch with different value at {:?}, {:?}, {:?}", alloc_id, offset, len);
+                *self = YieldRecordState::MadeProgress;
+            }
+        }
+    }
+
+    /// Mark a synchronization object as watched.
+    fn on_watch_sync(&mut self, sync: SyncObject) {
+        if let YieldRecordState::Recording {
+            watch_sync, record_iteration, ..
+        } = self {
+            let count = watch_sync.entry(sync).or_insert(0);
+            if *count < *record_iteration - 1 {
+                // Different content - assume progress.
+                log::trace!("Sync watch with different value at {:?}", sync);
+                *self = YieldRecordState::MadeProgress; 
+            } else {
+                *count = *record_iteration;
+            }
+        }
+    }
+
+    /// Returns true if the atomic variable is currently watched.
+    fn should_wake_atomic(&self, alloc_id: AllocId, offset: Size, len: Size) -> bool {
+        if let YieldRecordState::Recording {
+            watch_atomic, ..
+        } = self {
+            if let Some(range_map) = watch_atomic.get(&alloc_id) {
+                range_map.iter(offset, len).any(|(_, &watch)| watch != 0)
+            } else {
+                false
+            }
+        } else {
+            // First iteration yield, no wake metadata
+            // so only awaken after there are no enabled threads.
+            false
+        }
+    }
+
+    /// Returns true if the sync object is currently watched.
+    fn should_wake_sync(&self, sync: SyncObject) -> bool {
+        if let YieldRecordState::Recording {
+            watch_sync, ..
+        } = self {
+            if let Some(count) = watch_sync.get(&sync) {
+                *count != 0
+            } else {
+                false
+            }
+        } else {
+            // First iteration, no wake metadata.
+            false
+        }
+    }
+
+    /// Returns the number of yield iterations that have been executed.
+    fn get_iteration_count(&self) -> u32 {
+        if let YieldRecordState::Recording {
+            record_iteration, ..
+        } = self {
+            *record_iteration
+        } else {
+            0
+        }
+    }
+
+    fn should_watch(&self) -> bool {
+        if let YieldRecordState::Recording {
+            watch_atomic, watch_sync, ..
+        } = self {
+            // Should watch if either watch hash-set is non-empty
+            !watch_atomic.is_empty() || !watch_sync.is_empty()
+        } else {
+            false
+        }
+    }
+
+    /// Starts the next yield iteration
+    fn start_iteration(&mut self) {
```

Comment:
This is *only called by the scheduler*, in case when there was no "regular" thread to make progress with any more and we woke up a `DelayOnYield` thread, right? If so, please say that in the doc comment; it took me a while to realize this.